salt/_renderers/better_kubernetes.py

```diff
@@ -0,0 +1,104 @@
+"""A renderer for Kubernetes YAML manifests.
+
+Given a Kubernetes YAML file (which may be a stream of objects, i.e. YAML
+snippets separated by `---` lines), this will render a sequence of states
+(represented as an OrderedDict), mapping every such object to an invocation
+of our custom `object_[present|absent]` state function.
+
+To use it, add a shebang like `#!kubernetes` as the first line of your
+manifests SLS file. Optionally, you can use rendering pipelines (if templating
+is required), e.g. `#!jinja | kubernetes`.
+
+The shebang also supports passing options to this renderer, in the format
+`#!kubernetes argument1=value1&argument2=value2` (basically, a query string in
+the `application/x-www-form-urlencoded` format).
+The supported options are:
+- `kubeconfig`, the path to the kubeconfig file to use for communicating with
+  K8s API
+- `context`, the context from the kubeconfig to use
+- `absent`, a boolean to toggle which state function variant (`object_present`
+  or `object_absent`) to use (defaults to False)
+
+TODO: improve management of `kubeconfig` and `context` parameters, relying on
+      master configuration and sane defaults - look into `__opts__`
+"""
+from salt.exceptions import SaltRenderError
+from salt.ext import six
+from salt.utils.yaml import SaltYamlSafeLoader
+from salt.utils.odict import OrderedDict
+import yaml
+
+__virtualname__ = 'kubernetes'
+
+
+def __virtual__():
+    return __virtualname__
+
+
+def _step_name(obj, absent=False):
+    try:
+        namespace = obj.metadata.namespace
+    except AttributeError:
+        namespace = None
+
+    try:
+        name = obj.metadata.name
+    except AttributeError:
+        raise SaltRenderError('Object `metadata.name` must be set.')
+
+    if namespace is not None:
+        full_name = '{}/{}'.format(
+            namespace,
+            name,
+        )
+    else:
+        full_name = name
+
+    return "{verb} {api_version}/{kind} '{name}'".format(
+        verb='Remove' if absent else 'Apply',
+        api_version=obj.apiVersion,
+        kind=obj.kind,
+        name=full_name,
+    )
+
+
+def _step(manifest, kubeconfig=None, context=None, absent=False):
+    """Render a single Kubernetes object into a state 'step'."""
+    try:
+        assert __utils__['kubernetes.validate_manifest'](manifest)
+    except (AssertionError, ValueError) as exc:
+        raise SaltRenderError('Invalid manifest: {!s}'.format(exc))
+
+    obj = __utils__['kubernetes.convert_manifest_into_object'](manifest)
```

Comment:
This is already done by `validate_manifest`, maybe this could return this as well? Not sure what such API would look like though.

Reply:
Yeah I wondered... Since `__utils__` are only callable from other Python modules, I'd say we can safely change the API for validation to something along those lines:

```python
try:
    obj == __utils__['kubernetes.convert_manifest_into_object'](manifest)
except ValueError as exc:
    raise SaltRenderError('Invalid manifest: {!s}'...)

conversion_valid, diff = __utils__['kubernetes.validate_conversion'](obj, manifest)
if not conversion_valid:
    raise SaltRenderError(
        'Conversion of manifest is invalid. Removed fields: {}. Changed fields: {}.'.format(
            ', '.join(diff.removed()), ', '.join(diff.changed())
        )
    )
```